mpz-core/src/aes.rs

```diff
@@ -42,4 +42,89 @@ impl FixedKeyAes {
 
         Block::from(out)
     }
+
+    /// Correlation-robust hash function for 128-bit inputs
+    /// (cf. <https://eprint.iacr.org/2019/074>, §7.2).
+    /// The function computes `π(x) ⊕ x`.
+    /// π(x) = AES(fixedkey,x)
+    #[inline]
+    pub fn cr(&self, block: Block) -> Block {
+        let mut x = GenericArray::from(block);
+        self.aes.encrypt_block(&mut x);
+        Block::from(x) ^ block
+    }
+
+    /// Circular correlation-robust hash function
+    /// (cf.<https://eprint.iacr.org/2019/074>, §7.3).
+    ///
+    /// The function computes `H(sigma(x))`, where `H` is a correlation-robust hash
+    /// function and `sigma( x0 || x1 ) = (x0 xor x1) || x1`.
+    #[inline]
+    pub fn ccr(&self, block: Block) -> Block {
+        self.cr(Block::sigma(block))
+    }
+}
+
+/// A wrapper of aes, only for encryption.
+#[derive(Clone)]
+pub struct AesEncryptor(Aes128Enc);
+
+impl AesEncryptor {
+    /// AES_BLOCK_SIZE
+    pub const AES_BLOCK_SIZE: usize = 8;
+
+    /// Initiate an AesEncryptor instance with key.
+    #[inline(always)]
+    pub fn new(key: Block) -> Self {
+        let _key: [u8; 16] = key.into();
+        AesEncryptor(Aes128Enc::new_from_slice(&_key).unwrap())
+    }
+
+    /// Encrypt a block.
+    #[inline(always)]
+    pub fn encrypt_block(&self, blk: Block) -> Block {
+        let mut ctxt = GenericArray::from(blk);
+        self.0.encrypt_block(&mut ctxt);
+        Block::from(ctxt)
+    }
+
+    /// Encrypt many blocks.
+    #[inline(always)]
+    pub fn encrypt_many_blocks<const N: usize>(&self, blks: [Block; N]) -> [Block; N] {
+        let mut ctxt = [Block::default(); N];
+        for i in 0..N {
+            ctxt[i] = self.encrypt_block(blks[i]);
+        }
+        ctxt
+    }
+
+    /// Encrypt many blocks with many keys.
+    #[inline(always)]
+    pub fn para_encrypt<const NK: usize, const NM: usize>(keys: &[Self; NK], blks: &mut [Block]) {
+        let ptr = blks.as_mut_ptr() as *mut [Block; NM];
+        for i in 0..NK {
+            let ctxt = unsafe { &mut *ptr.add(i) };
+            *ctxt = keys[i].encrypt_many_blocks(*ctxt);
+        }
```

Comment:
This sort of pointer manipulation best be avoided.  I'm not sure I understand exactly what is going on here. What happens if `blks.len() < NK` or `blks.len() < NM`? This would lead to undefined behavior but is not warned in the docs nor is the function marked unsafe.

We should at least assert, but iterators can also be used as well:

```suggestion
        assert!(blks.len() >= NK + NM);
        let mut ctxt = [Block::default(); NM];
        keys.iter().enumerate().for_each(|(i, key)| {
            ctxt.copy_from_slice(&blks[i..NM + i]);
            blks[i..NM + i].copy_from_slice(&key.encrypt_many_blocks(ctxt))
        })
```

The above suggestion does not pass the unit test, so it may not be exactly what you had in mind but my point is to illustrate similar without using unsafe.

Reply:
updated without using `unsafe`